### ISSUE TITLE
duplicate entry

### ISSUE THREAD

**Issue by ferrored** [OpenSIPS/opensips#764]
I'm using opensips 1.11.5-tls on a ubuntu 14.04-tls server. The module usrloc if configure as:

modparam("usrloc", "db_mode", 2) 
modparam("usrloc", "desc_time_order", 1) 
modparam("usrloc", "matching_mode", 0)  #contact only

Sometime I get the following error:
CRITICAL:db_mysql:wrapper_single_mysql_stmt_execute: driver error (1062): Duplicate entry 'name.surname-domain.it-sip:name.surname@IP-' for key 'account_contact_idx'
Jan 19 14:33:06 proxysip /sbin/opensips[107504]: ERROR:usrloc:db_update_ucontact: updating database failed
Jan 19 14:33:06 proxysip /sbin/opensips[107504]: ERROR:usrloc:wb_timer: updating contact in db failed

I'm incapable of reproducing it.
Thank you in advance

Piero

**Comment by bogdan-iancu** [comment]
@ferrored , I assume your opensips instance is the only one writing to that table ?

**Comment by ferrored** [comment]
yes, it is

**Comment by bogdan-iancu** [comment]
did the error happen after some restarts or so ?

**Comment by ferrored** [comment]
Not necessarily after a reboot.
I noticed that sometimes it occurs when there is already a record in the table location, a phone is  disconnected from the network  (the un-register message is NOT sent), and when it is restarted the Aor is overwrote correctly in the memory, but when opensips try to write in the db the the errors appear

**Comment by ferrored** [comment]
On version 1.11.6-tls the errors are a bit differents:

CRITICAL:db_mysql:wrapper_single_mysql_real_query: driver error (1062): Duplicate entry 'prove-domain-sip:prove@IP-c83b28bb-5cb83d00-d95e' for key 'account_contact_idx'
Jan 22 10:44:21 proxysiptest /sbin/opensips[21541]: ERROR:core:db_do_update: error while submitting query

**Comment by ferrored** [comment]
[polycom-con-registrazioni-multiple.txt](https://github.com/OpenSIPS/opensips/files/117112/polycom-con-registrazioni-multiple.txt)

For some reasons that we don't understand, there are two records in the DB referred to the some Aor (see the attachment), and it seems that opensips isn't capable of deleting the expired registration.
If we try do delete the Aor (opensipsctl ul rm name.surname), the expired Aor isn't erased

**Comment by bogdan-iancu** [comment]
@ferrored , could you confirm which is the unique key for your mysql table ?

**Comment by ferrored** [comment]
location table
primary key: id,
unique key "account_contact_idx": (username,domain,contact,callid)

**Comment by bogdan-iancu** [comment]
If so, how comes the error, as the callid's are different in the 2 records you have. 

Could you match what opensips has in memory (opensipsctl ul show) versus what is in DB for that AOR ? (at the moment of error)

**Comment by ferrored** [comment]
opensipsctl ul show name.surname


Contact:: 
<sip:name.surname@IP>;q=0.9;expires=1367;flags=0x0;cflags=0x0;socket=<udp:149.132.2.72:5060>;callid=<f10b8916-855dc26-6dbb9b96@IP>;methods=0x1F7F;user_agent=<PolycomSoundPointIP-SPIP_450-UA/4.0.7.2514_0004f2483f2d>


there is only one Aor...


Piero

**Comment by ferrored** [comment]
Hi Bogdan,

if I use the "f" parameter in the function "save()" , save("vf"), there are no more error messages in the log, but in table location there are still 2 records, one expired that should be erased by opensips,
while in memory there is only the correct Aor

memory:
opensipsctl ul show name.surname
Contact:: <sip:name.gervasoni@IP>;q=0.9;expires=664;flags=0x0;cflags=0x0;socket=<udp:IP-PROXY-SIP:5060>;callid=<a21061c0-1b82fde7-c60381e8@IP>;methods=0x1F7F;user_agent=<PolycomSoundPointIP-SPIP_450-UA/4.0.7.2514_0004f2435806>

table location:
see attachment
[duplicate-error.txt](https://github.com/OpenSIPS/opensips/files/122915/duplicate-error.txt)




Piero

**Comment by ferrored** [comment]
Hi Bogdan,

if I restart the phone, associated to the identity name.surname, I found three records in the table location
[duplicate-error.txt](https://github.com/OpenSIPS/opensips/files/122956/duplicate-error.txt)

while in memory:
opensipsctl ul show name.surname
Contact:: <sip:name.surname@IP-PHONE>;q=0.9;expires=3414;flags=0x0;cflags=0x0;socket=<udp:IP-PROXY-SIP:5060>;callid=<6a900e81-38a33250-a74b641d@149.132.28.245>;methods=0x1F7F;user_agent=<PolycomSoundPointIP-SPIP_450-UA/4.0.7.2514_0004f2435806>

Piero

**Comment by ferrored** [comment]
Hi Bogdan,

I can reproduce the duplicate error.
Set usrloc time_interval=15 and execute
>sipp -sf duplicate-scenario.xml -inf register-scenario.csv 149.132.5.117 -m 100
[duplicate-scenario.xml.txt](https://github.com/OpenSIPS/opensips/files/132062/duplicate-scenario.xml.txt)
[Uploading register-scenario.csv.txt…]()

You should find that memory and DB are not synchronized

**Comment by bogdan-iancu** [comment]
Hi @ferrored ,

Thanks for the additional info, let me give it a try.

**Comment by bogdan-iancu** [comment]
@ferrored , I did not get the duplicate error, but I got the "leak" in the DB. Some records do stay there even if mem is cleared out.
Let me dig into this and find the reason.

**Comment by ferrored** [comment]
To get get the duplicate error you should send other two register messages for the user

**Comment by l7s** [comment]
Hi, I can confirm getting identical errors in our v.1.11.1 

```
Feb 15 16:46:12 uk1 /usr/sbin/opensips[19041]: CRITICAL:db_mysql:wrapper_single_mysql_real_query: driver error (1062): Duplicate entry '599802--sip:599802@92.40.248.209:41468;transport=UDP;rinstance=1' for key 'account_contact_idx'
Feb 15 16:46:12 uk1 /usr/sbin/opensips[19041]: CRITICAL:db_mysql:wrapper_single_mysql_real_query: driver error (1062): Duplicate entry '599802--sip:599802@92.40.248.209:41468;transport=UDP;rinstance=1' for key 'account_contact_idx'
Feb 15 17:17:29 uk1 /usr/sbin/opensips[19041]: CRITICAL:db_mysql:wrapper_single_mysql_real_query: driver error (1062): Duplicate entry '590981--sip:590981@31.4.229.32:35276;transport=UDP;rinstance=b84' for key 'account_contact_idx'
Feb 15 17:17:29 uk1 /usr/sbin/opensips[19041]: CRITICAL:db_mysql:wrapper_single_mysql_real_query: driver error (1062): Duplicate entry '590981--sip:590981@31.4.229.32:35276;transport=UDP;rinstance=b84' for key 'account_contact_idx'
```
my impression is it started since we switched to db_virtual in FAILOVER mode with Primary and Backup. Unable to reproduce this consistently in our Dev environment, however seeing 20-50 above log entries every day.

Regards,
Chris

**Comment by bogdan-iancu** [comment]
@ferrored , please test the fix I just pushed and let me know if it solves your problem.
Thanks, Bogdan

**Comment by ferrored** [comment]
I applied the patch in my dev environment and it  works.

Thank you very much

Piero

**Comment by bogdan-iancu** [comment]
Hi @ferrored , thanks for the confirmation - I just pushed the fix on 2.1 branch too (2.2 is not affected as it has a completely different way of operating with the contacts as DB level)
